Feature: Raft node metrics

tdietert opened the pull request:
A per discussion earlier today, this PR adds individual raft node metrics using the `monad-metrics` library. By design, the library exposes a `MonadMetric` typeclass that ultimately needs IO as the base monad. This typeclass allows for the addition of various types of metrics:
- `Counter`: a key value store mapping `Text` values to their "counts"
- `Distribution`: a running statistical analysis of values recorded at arbitrary times
- `Label`:  a key value store mapping `Text` values to other `Text` values

Ultimately, these metrics will be read by client programs that wish to enforce that certain node-level properties hold under specific conditions. This PR should decide upon an API for querying these node metrics. There are several ways to do this, each with tradeoffs. For brevity, I will simply list the means instead of also discussing their tradeoffs:

- Forking an thread that handle REST style requests, returning a JSON value of the node metrics to the issuer.
- Adding a new `ClientRequest` value that all nodes service as a request and respond to the issuer through use of typeclass functions similar to or exactly as `RaftSendClient`, `RaftRecvClient`, and `RaftClientSend`, and `RaftClientRecv`.

The initial commit of this PR achieves the following:
- Adds a `MonadMetric` instance to the `RaftT` monad 
- Initializes a `Metric` value and adding it to the `RaftEnv` for use within `RaftT` computations
- Increments a "Invalid Command" counter that counts the number of invalid commands seen by a node.
- Computes a distributed over the amount of time that each `handleAction` function call makes (note: this inhibits performance a little, and is added for demonstration purposes).

tdietert commented:
It would be really great if we could do the following things in this PR:
- Add the option to fork a metrics server or not (if `configMetricsPort` is `Nothing`, no metrics server)
- Add an (optional) cmd line option for `raft-example` to specify the port for the metrics server to run on
- Add config sanitization, reporting an error if invalid config values are specified (invalid node ids, invalid metrics port numbers, etc)

faineance commented on stack.yaml:
```diff
@@ -12,3 +12,6 @@ extra-deps:
 - tasty-dejafu-1.2.1.0
+- monad-metrics-0.2.1.2
+- ekg-0.4.0.15
+- ekg-json-0.1.0.6
```
Not sure I see ekg-json being used?

tdietert replied:
:+1: I think it may be left over from a dependency I added and then removed during the development process of this feature. Just pushed a commit to remove it.

tdietert commented:
> > Computes a distributed over the amount of time that each handleAction function call makes (note: this inhibits performance a little, and is added for demonstration purposes).
> 
> I don't see how this pr achieves this? maybe I'm missing something.

Indeed, it no longer does. I was worried about the performance implications about timing the `handleEvent` function when dozens of events are handled per second, so I removed it without updating the PR description.

Where I added it:
https://github.com/adjoint-io/raft/pull/122/commits/5922857f461a66588cdabc6481622ab6b3bd264b#diff-31544c1a61d759321301686ad413aaeeR345

Where I deleted it:
https://github.com/adjoint-io/raft/pull/122/commits/0e540bcff27a15a185c2a277df8211c885f8cc5d#diff-31544c1a61d759321301686ad413aaeeL358

It may be worth having, however I am unsure of the performance implications and how much it may affect the processing of the list of actions generated by the `handleEvent` function.

tdietert commented on src/Raft/Metrics.hs:
```diff
@@ -0,0 +1,64 @@
+  deriving Show
+
+incrRaftNodeCounter :: (MonadIO m, Metrics.MonadMetrics m) => RaftNodeCounter -> m ()
+incrRaftNodeCounter = Metrics.increment . show
```
:ok_man:

tdietert commented:
which, unfortunately, it is not :thinking: :grimacing: 
I'll dig into the logs...

tdietert commented:
Are there any more simple metrics that we'd like to include in this PR?
I think as @faineance mentioned, it would be nice to know the node state: whether the node is currently a follower, candidate, or leader. I will add that now, but at the same time are there any other metrics of interest at the moment?
